using_restricted_array.rb

```diff
@@ -18,36 +19,88 @@ def length(array)
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    print "#{array[i]}"
+    if array[i+1] == nil
+      print "\n"
+    else
+      print ', '
+    end
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until i == length
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  max = array[0]
+  until i == length
+    if array[i] > max
+      max = array[i]
+    end
+    i += 1
+  end
+  max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  min = array[0]
+  until i == length
+    if array[i] < min
+      min = array[i]
+    end
+    i += 1
+  end
+  min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  a = 0
+  b = length - 1
+  until a >= b
+    array[a] = array[a] + array[b]
+    array[b] = array[a] - array[b]
+    array[a] = array[a] - array[b]
+    a += 1
+    b -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  min = 0
```

Comment:
Please take a look at the updated binary_search slide. Here's the considerations I was missing in the algorithm on the slide before the update: (1) what if there is only one element in the array and it is the value_to_find? In this case, it used to return false. (To fix this, add an additional if statement after the loop ends, to check if array[min] is the value_to_find.) (2) Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update max to be mid -1, if value_to_find is less than mid. And update min = mid+1, if value_to_find is greater than mid.)